File: src/data_rover/validation/validators/facets.py
from __future__ import annotations

import re

from ..issue import Issue, Severity
from ..scope import Scope


class FacetsValidator:
    def validate(self, model, scope: Scope) -> list[Issue]:
        issues: list[Issue] = []
        mm = model.metamodel
        for el in model.elements.values():
            if not scope.includes(el.id):
                continue
            defs = {p.name: p for p in mm.effective_element_properties(el.type_name)}
            for name, value in el.properties.items():
                pdef = defs.get(name)
                if pdef is None or value is None:
                    continue
                values = value if isinstance(value, list) else [value]
                for item in values:
                    issues.extend(self._check(el.id, name, pdef, item))
        return issues

    def _check(self, eid, name, pdef, item) -> list[Issue]:
        out: list[Issue] = []
        if isinstance(item, (int, float)) and not isinstance(item, bool):
            if pdef.min is not None and item < pdef.min:
                out.append(
                    Issue(Severity.ERROR, f"{name}: {item} below min {pdef.min}", [eid])
                )
            if pdef.max is not None and item > pdef.max:
                out.append(
                    Issue(Severity.ERROR, f"{name}: {item} above max {pdef.max}", [eid])
                )
        if isinstance(item, str):
            if pdef.pattern is not None and not re.fullmatch(pdef.pattern, item):
                out.append(
                    Issue(
                        Severity.ERROR,
                        f"{name}: {item!r} does not match pattern {pdef.pattern!r}",
                        [eid],
                    )
                )
            if pdef.max_length is not None and len(item) > pdef.max_length:
                out.append(
                    Issue(
                        Severity.ERROR,
                        f"{name}: length {len(item)} exceeds "
                        f"max_length {pdef.max_length}",
                        [eid],
                    )
                )
        return out

File: engine/src/validation/validators/facets.ts
import type { Metamodel } from '../../metamodel/metamodel.ts';
import type { PropertyDef } from '../../metamodel/types.ts';
import { getProp, type ElementRec, type Props, type RelRec } from '../../model/records.ts';
import { pyFloatRepr } from '../../value/float-repr.ts';
import { pyRepr } from '../../value/repr.ts';
import { PyFloat, type Value } from '../../value/types.ts';
import { errorIssue } from '../issue.ts';
import type { Run, Validator } from '../pipeline.ts';
import { pyStrNumber } from '../values.ts';

/** Python's `len` of a string: code points, a lone surrogate counting one. */
function pyLen(s: string): number {
	let n = s.length;
	for (let i = 0; i < s.length - 1; i++) {
		const unit = s.charCodeAt(i);
		if (unit >= 0xd800 && unit <= 0xdbff) {
			const next = s.charCodeAt(i + 1);
			if (next >= 0xdc00 && next <= 0xdfff) {
				n--;
				i++;
			}
		}
	}
	return n;
}

/**
 * `min` / `max` on every int and float value, `pattern` (a full match) and
 * `max_length` on every string value, whatever the property's datatype. A
 * bound is a float, and renders as one.
 */
export class Facets implements Validator {
	readonly checkName = 'facets';
	private readonly mm: Metamodel;
	private readonly elementDefs = new Map<string, readonly PropertyDef[]>();
	private readonly relationshipDefs = new Map<string, readonly PropertyDef[]>();

	constructor(mm: Metamodel) {
		this.mm = mm;
	}

	/** The type's effective properties that carry a facet; most carry none. */
	private defs(typeName: string, ofElement: boolean): readonly PropertyDef[] {
		const cache = ofElement ? this.elementDefs : this.relationshipDefs;
		let defs = cache.get(typeName);
		if (defs === undefined) {
			const props = ofElement
				? this.mm.effectiveElementProperties(typeName)
				: this.mm.effectiveRelationshipProperties(typeName);
			defs = props.filter(
				(p) => p.min !== null || p.max !== null || p.pattern !== null || p.max_length !== null
			);
			cache.set(typeName, defs);
		}
		return defs;
	}

	validateElement(run: Run, el: ElementRec): void {
		const defs = this.defs(el.typeName, true);
		if (defs.length > 0) this.checkProps(run, el.id, defs, el.props);
	}

	validateRelationship(run: Run, rel: RelRec): void {
		const defs = this.defs(rel.typeName, false);
		if (defs.length > 0) this.checkProps(run, rel.id, defs, rel.props);
	}

	private checkProps(run: Run, ownerId: string, defs: readonly PropertyDef[], props: Props): void {
		for (const def of defs) {
			const value = getProp(props, def.name);
			if (value === undefined || value === null) continue;
			for (const item of Array.isArray(value) ? value : [value])
				this.check(run, ownerId, def, item);
		}
	}

	private check(run: Run, ownerId: string, def: PropertyDef, item: Value): void {
		const { name } = def;
		if (typeof item === 'number' || typeof item === 'bigint' || item instanceof PyFloat) {
			const x = item instanceof PyFloat ? item.value : item;
			if (def.min !== null && x < def.min) {
				run.out.push(
					errorIssue(`${name}: ${pyStrNumber(item)} below min ${pyFloatRepr(def.min)}`, [ownerId])
				);
			}
			if (def.max !== null && x > def.max) {
				run.out.push(
					errorIssue(`${name}: ${pyStrNumber(item)} above max ${pyFloatRepr(def.max)}`, [ownerId])
				);
			}
		}
		if (typeof item === 'string') {
			if (def.pattern !== null && !run.patterns.fullmatch(def.pattern, item)) {
				run.out.push(
					errorIssue(`${name}: ${pyRepr(item)} does not match pattern ${pyRepr(def.pattern)}`, [
						ownerId
					])
				);
			}
			if (def.max_length !== null) {
				const length = pyLen(item);
				if (length > def.max_length) {
					run.out.push(
						errorIssue(`${name}: length ${length} exceeds max_length ${def.max_length}`, [ownerId])
					);
				}
			}
		}
	}
}
